server/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -1591,6 +1610,8 @@ private Node stop() {
     // In this case the process will be terminated even if the first call to close() has not finished yet.
     @Override
     public synchronized void close() throws IOException {
+        Optional.ofNullable(terminationHandler.get()).ifPresent(TerminationHandler::handleTermination);
```

Comment:
Why call this at close?  The javadoc in TerminationHandler indicates `handleTermination` is a result of a signal from the OS.  That handling occurs in [Elasticsearch.java](https://github.com/elastic/elasticsearch/blob/cbcbd01d5e78026d12b9ea29834f39e45def9a5a/server/src/main/java/org/elasticsearch/bootstrap/Elasticsearch.java#L442).

Reply:
Plumbing, basically - we don't have a `PluginsService` in `Elasticsearch.java` to get the termination handlers. We could potentially lift the construction of `PluginsService` out of `Node` into `Elasticsearch`, or expose the termination handler as a separate method on `Node` and call it separately if you think that would be better.